util/src/main/java/com/google/cloud/hadoop/util/RetryHttpInitializer.java

```diff
@@ -154,7 +160,8 @@ public CredentialOrBackoffResponseHandler() {
       errorCodeHandler.setBackOffRequired(
           response ->
               BASE_HTTP_BACKOFF_REQUIRED.isRequired(response)
-                  || response.getStatusCode() == HTTP_SC_TOO_MANY_REQUESTS);
+                  || response.getStatusCode() == HTTP_SC_TOO_MANY_REQUESTS
+                  || response.getStatusCode() == HTTP_REQUEST_TIMEOUT);
```

Comment:
I think that `google-cloud-storage` library added retries only for idempotent requests (`GET`, `HEAD`, etc), maybe we need to check with them and clarify if/why this additional check necessary?

cc @benwhitehead

Reply:
This is correct. Not all GCS operations neatly fit into the standard HTTP idempotency conventions, in some cases extra parameters need to be taken into account to determine if a request is idempotent.

These docs https://cloud.google.com/storage/docs/retry-strategy#idempotency give a good overview of some of the specifics.

Reply:
Good point Igor, ideally these should be handled by libraries themselves.
1. Let me know if I understand correctly. We use `google-api-services-storage` instead of `google-cloud-storage` from https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcsio/pom.xml#L70.

2. Thanks for link @BenWhitehead, on same page, I can see under Client Libraries -> Java, 408 is actually retried by default. Under code snippet, I see `com.google.cloud.storage.Storage` as class while we use `com.google.api.services.storage.Storage` (code link below), is the documentation still valid?

3. @medb in the above documentation, it is mentioned that we can override default settings which we do but bit differently from what is mentioned in the sample code, we use HttpRequestInitializer instead, this may be the reason why we are not benefitting from default retry settings? https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java#L364 cc: @BenWhitehead

Reply:
Correct `google-api-services-storage` is the Discovery doc based automatically generated JSON client (aka Apiary client) (`google-cloud-storage` also uses this for the majority of it's RPCs) which doesn't have any automatic retry AFAIK, something like a request interceptor or request initializer is what would need to be leveraged to have the Apiary client perform multiple attempts. In `google-cloud-storage` we handle retries at a higher level as some of our RPCs are out of band of the apiary client.

Any links to code from `google-cloud-storage` are intended to be purely informational to show how that client has approached retry handling for some cases. The https://cloud.google.com/storage/docs/retry-strategy#idempotency doc is the general public which serves as a guide for anyone doing client level work against gcs (including the `google-cloud-storage` team as well as other customers).